### DISCUSSION TITLE
Wayland additional vblank-timing of 5percent instead of 15percent

### DISCUSSION THREAD

**Discussion by stonix015** [mpv-player/mpv#15103] [General]
Hey @Dudemanguy ,

Firstly: I want to thank you so much for maintaining the mpv player.
It is awsome! And there are reasons for me to use it against other players, even now, where i went into trouble and could solve the problem!

**Problem:**
The situation was that i had **vsync troubles** when watching 24fps films on a 24Hz TV (everything set to 24Hz in system). My frame timing were lower than 10ms, but i had missed frames and the playback was occasionally stuttering in a noticeable manner.
I've read a lot research on how things go between the video-file and the final playback on the display and where a mistake could happen. So i figured out it **should be a sync problem, not a performance problem**.

**Solution:**
I've found the following thread: https://github.com/mpv-player/mpv/issues/9504 in which you state that a magical empirical number of 5 percent onto the vblank timing could resolve an issue on wayland.
So now i advanced this number to 15 percent:
`video/out/wayland-common.c` -> `vblank_time += 0.15 * vblank_time; ` instead of `vblank_time += 0.05 * vblank_time; `
Maybe the magical number of 5 percent is too less in some applications. 
And as i spend a lot of time into that, i wanted to notice to you that experience.

**System:**
The problem appeared in a fully up-to-date **Arch Linux distro with linux 6.11** and **Plasma-wayland**.
**Linux-firmware** package is installed as also **mesa-drivers**, **intel-media-driver** and all other possible needed goodies.
The system is an **Intel i5-n305** Alder Lake with its integrated UHD graphics card (with 32EUs). 
It is running on **8GB DDR5-SODIMM**, the **GPU gets 256MB**, higher values in BIOS made the situation worse for me.

**Other Steps:**
What i tried and what didnt fix the issue was:
X11 with vulkan, X11 with opengl, Wayland with vulkan, Wayland with opengl, and everything once again with the xperimental Xe Iris driver set as kernel-option.

If you need some more information, stay in contact.
Maybe i can help here for a further issue of mpv.

Greetings,
Lukas

**Comment by Dudemanguy** [comment]
X11 shouldn't have this problem. Or at least if it had similar symptoms, the root cause should be something completely different.

As a workaround for now, you can use `--wayland-disable-vsync` + `--opengl-swapinterval=1` or `--vulkan-swap-mode=fifo` to do the vsync block completely on the drivers. That should work fine but it will have the downside of mpv freezing if you occlude the window in some way. That is why we don't default that. In the future for wayland, the proper fix is https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/26150 and along with this https://github.com/mpv-player/mpv/pull/15056 while using vulkan. Kwin would need to support these protocols too, but they will almost certainly do it. This let us use fifo without any downsides and there should be no need for these guess hacks.

I could make that random timing configurable but I would really rather not. Users shouldn't have to mess with this.

**Reply by stonix015** [reply]
I was also wondering why a now get the same problems with X11 via modesetting driver.
with my ThinkPad (8th gen) on X11 everything runs flawlessy, but with xf86-video-intel driver, which is now outdated and should not be used with 12th gen alder lake GPUs. (I already ran into trouble with that, no it really doesnt work).

So my guess is, that the modesetting driver plays a role on X11.

I will try the fifo / swapinterval setting out. This sounds interesting.
Because with `wayland-disable-vsync` but also with `display-resample` the video speed was a complete mess. (even in the 24hz TV <-> 23,98fps setting)

Shouldnt i get into tearing-troubles with that?

And: as you mention the proper fix for that: How long will that "future" be from now?

**Reply by llyyr** [reply]
> And: as you mention the proper fix for that: How long will that "future" be from now?

Next Mesa and kwin feature release, probably? The protocol is merged so the rest shouldn't take long considering there's implementations for these things already, they just need to be reviewed and merged.

**Reply by Dudemanguy** [reply]
If you set `--wayland-disable-vsync` setting the swapinterval for opengl or fifo is not optional. Like the name implies, it disables our own vsync heuristic so you will need to rely on the driver instead. This is also only viable for `vo_gpu` or `vo_gpu_next`. If you happen to use `vo_dmabuf_wayland`, it needs the internal vsync to work correctly.

>Shouldnt i get into tearing-troubles with that?

No it should be impossible for mpv to tear on wayland unless you use `--vulkan-swap-mode=immediate` on a compositor that supports that and even then it would only happen on fullscreen.

>And: as you mention the proper fix for that: How long will that "future" be from now?

I'd expect a month or two before everything lands into the next releases.

**Reply by stonix015** [reply]
Thank you very much.
From all ive read, i will test the setting as following:

vo=gpu-next
api=vulkan
context=waylandvk
wayland-disable-vsync=yes
video-sync=display-resample
vulkan-swap-mode=fifo
vulkan-queue-count=8 (does this have an advantage from now, if my gpu has troubles with a few single frames due to HQ-settings?)

**Reply by Dudemanguy** [reply]
I wouldn't expect vulkan-queue-count to make any real difference here.

**Reply by stonix015** [reply]
Hey, i'd just like to notice that ive got further:
On X11, and on wayland, nothing you (both) noticed helped. No mode did the trick, even my 15percent-code seemed to had some jitters.

But i ended up no in DRM mode unter tty without any graphical environment.
This is not only just fine for me, but an advance! Because i've got now the dedicated media player i wanted. You feel that mpv has full control of the video drivers, the picture is far more fluid, and the frame generation of the projector does a noticeable better job. 
VSync-Jitter is now absolutely nailed a 0,0004 (against 0,05 and lots of variation as a slave in graphical environments).

Next step is to get 10bit colors to work, because these do only work with FHD, but not UHD. I think after a few tests with my HDR pc monitor and my static computer (RX7900GRE) i figured out that the HDMI cable and the DP-HDMI adapter as the bottleneck.
Bought a 8k60 cable now and awaiting for it.

**Just a last question because i am curious:** From the DRM perspective, are there any other settings in the mpv.conf i have to set to get 10bit colors to work without any intermediated 8bit calculation steps in the processing? (With HDR video material)
I do not mean activating HDR mode (i do not want that at the projector, because it has some disadvantages). I want to stick with the tone-mappings i configured. I just want to use a full 10 bit color processing for the details.

**Reply by llyyr** [reply]
>  On X11, and on wayland, nothing you (both) noticed helped. No mode did the trick, even my 15percent-code seemed to had some jitters.

The code you're adjusting has no effect on anything related to x11, so if you have the issue on both x11 and wayland it's something entirely unrelated to your findings.

**Reply by stonix015** [reply]
> > On X11, and on wayland, nothing you (both) noticed helped. No mode did the trick, even my 15percent-code seemed to had some jitters.
> 
> The code you're adjusting has no effect on anything related to x11, so if you have the issue on both x11 and wayland it's something entirely unrelated to your findings.

Understood that. What i meant was: Nothing helped under Graphical environments. But in TTY via DRM everything runs like from a dedicated blu ray player. This helped a lot und definately will be the final solution. 

I now have to start the pc, Type in: mpv Films/Sci-fi/BLABLABLABLA und enter. This is comfortable, it only shocks a few people who dont know anything about computers and are like "Hackerman" :D

**Comment by llyyr** [comment]
Curious if you have the issue with vo=dmabuf-wayland as well or not

**Reply by stonix015** [reply]
I will try that. Next film will be in probably 2 hrs :D (one film per evening here, home theatre)

i will reinstall the non-custom mpv from the official reps for that testing.

Greetings!
You guys are really helping in a really fast manner.
Firstly i was curious if i'd open a case for that, i wanted to resolve the issue bymyself. But i have to be honest that i was on my limit now. 
The last week was completely testing every day for 5 hrs.... and the enjoyment of watching movies was completely destroyed.
